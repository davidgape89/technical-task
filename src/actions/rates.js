import axios from 'axios';

// Export for testing
export const ratesUrl = `https://api.exchangeratesapi.io/latest`;

export const setBaseCurrency = (base) => {
  return (dispatch) => {
    dispatch({type: 'SET_BASE', base});
    dispatch(requestRates());
  }
}

export const requestRates = () => {
  return (dispatch, getState) => {
    const {rates: {base}} = getState();
    
    return axios.get(`${ratesUrl}?base=${base}`)
      .then((response) => {
        let rates = response.data.rates;
        rates[base] = 1;
        dispatch(setRates(rates));
      })
      .catch((error) => 
        console.error('There was a problem fetching the rates'));
  }
}

export const setRates = (rates) => ({
  type: 'SET_RATES',
  rates
});