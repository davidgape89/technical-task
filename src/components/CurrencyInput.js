import React from 'react';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputAdornment from '@material-ui/core/InputAdornment';

const CURRENCY_REGEX = /^([0-9]{1,3},([0-9]{3},)*[0-9]{3}|[0-9]+).?([0-9]{0,2})$/;

export const CurrencyInput = ({
  label = 'Amount',
  value,
  placeholder,
  symbol,
  onChange,
}) => {
  const formatNumber = (number) => {
    number = number.toString();

    if (number && parseFloat(number) < Number.MAX_SAFE_INTEGER) {
      // Format integer part
      number = number.replace(/\d+/, integerPart => parseFloat(integerPart).toLocaleString('en-US'));
      // Take only two decimals
      return number.match(/[\d\,]*\.?\d{0,2}/)[0];
    }

    return '';
  };

  const onAmountChange = ({ target: { value } }) => {
    if (value.match(CURRENCY_REGEX)) {
      value = value.match(CURRENCY_REGEX)[0];
      value = value.replace(/\,/g, '');
      if (parseFloat(value) < Number.MAX_SAFE_INTEGER) {
        onChange(value);
      }
    } else if (value === '') {
      onChange('');
    }
  };

  return (
    <FormControl fullWidth>
      <InputLabel htmlFor="currency-input">{label}</InputLabel>
      <Input
        id="currency-input"
        variant="outlined"
        placeholder={placeholder}
        value={formatNumber(value)}
        onChange={onAmountChange}
        startAdornment={<InputAdornment position="start">{symbol}</InputAdornment>}
      />
    </FormControl>
  );
};
