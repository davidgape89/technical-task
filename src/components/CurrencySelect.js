import React from 'react';
import propTypes from 'prop-types';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';
import MenuItem from '@material-ui/core/MenuItem';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormControl from '@material-ui/core/FormControl';

export const CurrencySelect = ({
  currencies,
  onChange,
  value,
}) => {
  const balance = Math.round(currencies[value] * 100) / 100;
  return (
    <FormControl fullWidth>
      <Select
        value={value}
        onChange={event => onChange(event.target.value)}
        input={
          <OutlinedInput labelWidth={0} />
        }
      >
        {Object.keys(currencies).map(currency => (
          <MenuItem
            key={currency}
            value={currency}
          >
            {currency}
          </MenuItem>
        ))}
      </Select>
      <FormHelperText>
Balance:
        {balance}
      </FormHelperText>
    </FormControl>
  );
};

CurrencySelect.propTypes = {
  currencies: propTypes.object,
  onChange: propTypes.func,
  value: propTypes.string,
};

