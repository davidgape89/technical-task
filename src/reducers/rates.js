export const defaultState = {
  base: 'EUR',
  rates: {}
}

export const ratesReducer = (state = defaultState, action) => {
  switch(action.type) {
    case 'SET_BASE':
      return {...state, base: action.base};
    case 'SET_RATES':
      return {...state, rates: action.rates};
    default:
      return state;
  }
} 