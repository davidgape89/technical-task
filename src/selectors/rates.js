export const exchangeToValue = (rates, fromCurrency, toCurrency, amount) => amount / rates[fromCurrency] * rates[toCurrency];

export const exchangeFromValue = (rates, fromCurrency, toCurrency, amount) => amount / rates[toCurrency] * rates[fromCurrency];
