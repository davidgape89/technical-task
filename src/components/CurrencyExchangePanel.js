import React from 'react';
import propTypes from 'prop-types';
import { CURRENCY_SYMBOLS } from '../constants';
import { exchangeToValue } from '../selectors/rates';

const CurrencyExchangePanel = ({
  rates,
  fromCurrency,
  toCurrency,
}) => {
  const result = exchangeToValue(rates, fromCurrency, toCurrency, 1);
  return (
    <span>
      {CURRENCY_SYMBOLS[fromCurrency]}
      1&nbsp;=&nbsp;
      {CURRENCY_SYMBOLS[toCurrency]}
      {Math.round(result * 10000) / 10000}
    </span>
  );
};

CurrencyExchangePanel.propTypes = {
  rates: propTypes.objectOf(propTypes.number).isRequired,
  fromCurrency: propTypes.string.isRequired,
  toCurrency: propTypes.string.isRequired,
};

export default CurrencyExchangePanel;
