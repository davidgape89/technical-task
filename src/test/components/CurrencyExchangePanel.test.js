import React from 'react';
import renderer from 'react-test-renderer';
import rates from '../fixtures/rates';
import {CurrencyExchangePanel} from '../../components/CurrencyExchangePanel';

describe('CurrencyExchangePanel - ', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = renderer.create(
      <CurrencyExchangePanel rates={rates.rates} 
                             fromCurrency="EUR"
                             toCurrency="USD"/>
    ).toJSON();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});