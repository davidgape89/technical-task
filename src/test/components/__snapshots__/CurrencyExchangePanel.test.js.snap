// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyExchangePanel -  renders correctly 1`] = `
<span>
  €
  1 = 
  $
  1.1354
</span>
`;
